site/specification/ORCv1.md

```diff
@@ -369,6 +392,174 @@ message Metadata {
 }
 ```
 
+# Column Encryption
+
+ORC as of Apache ORC 1.6 supports column encryption where the data and
+statistics of specific columns are encrypted on disk. Column
+encryption provides fine-grain column level security even when many
+users have access to the file itself. The encryption is transparent to
+the user and the writer only needs to define which columns and
+encryption keys to use. When reading an ORC file, if the user has
+access to the keys, they will get the real data. If they do not have
+the keys, they will get the masked data.
+
+Each encrypted column in each file will have a random local key
+generated for it. Thus, even though all of the decryption happens
+locally in the reader, a malicious user that stores the key only
+enables access that column in that file. The local keys are encrypted
+by the Hadoop or Ranger Key Management Server (KMS). The encrypted
```

Comment:
Is it supposed to work with Hadoop ecosystem only?

Reply:
No, it should work with anything that will encrypt local keys for us. In the Java side, look at the KeyProvider API in the Hadoop shims module. It was designed to work with either the cloud specific KMS's such as Amazon's or the on-prem Hadoop or Ranger KMS. 

Your point however, remains very relevant however. We should also encode how the local keys were encoded. It looks like we need another enum in the Key that looks like:

```
enum KeyProviderKind {
  UNKNOWN = 0;
  HADOOP_KMS = 1;
  AWS_KMS = 2;
  GCP_KMS = 3;
  AZURE_KEY_VAULT = 4;
}
```

Reply:
After further reflection, I moved this to the Encryption section, so it becomes a file level property. I don't think we need to support a use case where some of the keys are encrypted/decrypted by different KeyProviders.